docs(cn): translate content/docs/hooks-reference.md into Chinese (#173)
reactjs/zh-hans.reactjs.org: content/docs/hooks-reference.md

---
id: hooks-reference
title: Hook API 参考
permalink: docs/hooks-reference.html
prev: hooks-custom.html
next: hooks-faq.html
---

> *Hook* 是 React 16.8 的新特性。利用它，你无需编写 class 就能使用 state 和其他 React 特性。

Author: 修改的句尾多了 ```

> *Hook* 是 React 16.8 的新特性。利用它，你无需编写 class 就能使用 state 和其他 React 特性。

Author: 改动的句尾多了 ```。这句话在每篇 hook 文档开头都有，看看要统一成哪个版本的翻译。

> *Hook* 是 React 16.8 的新特性。利用它，你无需编写 class 就能使用 state 和其他 React 特性。

Reviewer: > 改动的句尾多了 ```。这句话在每篇 hook 文档开头都有，看看要统一成哪个版本的翻译。

可能是复制的时候多带了，改掉了。这个是要统一一下。因为其他的API文档里也有，我是跟我已经Merge的那篇统一了。

> *Hook* 是 React 16.8 的新特性。利用它，你无需编写 class 就能使用 state 和其他 React 特性。

Author: ```suggestion
*Hook* 是 React 16.8 的新增特性。它可以让你在不编写 class 的情况下使用 state 以及其他的 React 特性。
```

暂时统一成这种翻译了，跟其他 hook 文档保持一致。

> 本页面描述 React 内置的 Hook API。

Reviewer: > Hooks -> Hook ?
已改

如果你刚开始接触 Hook，那么可能需要先查阅 [Hook 概览](/docs/hooks-overview.html)。你也可以在 [Hooks FAQ](/docs/hooks-faq.html) 章节中获取有用的信息。

- [基础 Hook](#basic-hooks)
  - [`useState`](#usestate)
  - [`useEffect`](#useeffect)
  - [`useContext`](#usecontext)

> - [额外的 Hook](#additional-hooks)

Reviewer 2: ```suggestion
- [额外 Hook](#additional-hooks)
```

Author: 这里要不要和 API Reference/React 的翻译统一

Reviewer 2: 哪个？

Author: > 哪个？

[React 顶层 API](/reactjs/zh-hans.reactjs.org/blob/master/content/docs/reference-react.md#hooks-hooks)，这里面 `additional` 译为 `额外的`

Reviewer 2: 嗯，这个和那个保持一致吧~

  - [`useLayoutEffect`](#uselayouteffect)
  - [`useDebugValue`](#usedebugvalue)

## 基础 Hook {#basic-hooks}

### `useState` {#usestate}

```js
const [state, setState] = useState(initialState);
```

> 返回一个 state，以及更新 state 的函数。

Reviewer 2: 因为 useState 就是 state，所以直接这样说更直观，直译感觉有点奇怪

> 返回一个 state，以及更新 state 的函数。

Author: 这里和哪些 review 冲突？是需要统一还是？

> 返回一个 state，以及更新 state 的函数。

Reviewer 2: @Yuiham 我也是这么认为的

> 返回一个 state，以及更新 state 的函数。

Author: 那保留译文吧

> 在组件初始渲染期间，`useState` 会返回和它第一个参数（`initialState`）等值的 state（`state`）。

Author: 全文都用 `state`，唯独这里用 `状态`，感觉还是有点奇怪。

`setState` 函数用来更新 state。它接收新的 state 值，然后将组件排入重渲染队列。

```js
setState(newState);
```

在后续的组件重渲染过程中，`useState` 的第一个返回值总是最新的 state。

> 注意
>
> React 会确保 `setState` 函数的标识是稳定的，并且不会在组件重新渲染时发生变化。这就是为什么可以安全地从 `useEffect` 或 `useCallback` 的依赖列表中省略 `setState`。

#### 函数式的更新 {#functional-updates}

如果需要基于上一轮的 state 来计算新的 state，可以把一个接收旧值并返回新值的函数传给 `setState`。下面的计数器组件示例使用了 `setState` 的两种用法：

```js

> function Counter({initialCount}) {

Author: 我倾向于不对代码本身做本地化处理，不知道大家意见如何？

> function Counter({initialCount}) {

Reviewer 2: 如果从便于理解的角度，不翻译比较好

}
```

由于需通过上一次的计数来计算新的计数值，「+」「-」按钮采用函数式的 `setState` 去处理更新。而「Reset」按钮则使用普通形式的 `setState`，因为它总是将计数置回 0。

> 注意
>

> > 与 class 组件的 `setState` 方法不同，`useState` 返回的 `setState` 函数并不会自动合并两次更新时接收的对象。你可以用函数式的 `setState` 结合对象拓展语法来合并 state 对象。

Reviewer 3: > 你通过结合函数式更新器与对象展开语法来实现该行为。

后半句这样知不是更直接一点。

> > 与 class 组件的 `setState` 方法不同，`useState` 返回的 `setState` 函数并不会自动合并两次更新时接收的对象。你可以用函数式的 `setState` 结合对象拓展语法来合并 state 对象。

Author: > > 你通过结合函数式更新器与对象展开语法来实现该行为。
> 
> 后半句这样知不是更直接一点。

结合上下文来看，我还是觉得把 `function updater` 显示地翻译为 `setState` 更加方便理解。

});
```

#### 跳过 state 更新 {#bailing-out-of-a-state-update}

> 假如你在调用 State Hook 的更新函数时，给它传入当前的 state 值，那本次更新将被跳过，React 既不会渲染子组件，也不会执行 effect。（React 使用 [`Object.is` 比较算法](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/is#Description) 来比较 state。）

Author: > If you update a State Hook to the same value as the current state
这里翻译为 “更新 State Hook” ，我觉得有歧义

### `useEffect` {#useeffect}

```js
useEffect(didUpdate);
```

`useEffect` 接收一个包含命令式的、可能有副作用代码的函数。

在函数组件主体内（这里指在 React 渲染阶段）变更DOM、设置订阅、计时、记录日志，以及执行其他副作用等操作都是不可取的，因为这可能引发令人困扰的 bugs，破坏 UI 的一致性。

> 相反，应该使用 `useEffect` 来完成上述操作。传给 `useEffect` 的函数会在组件渲染完成后执行。你可以把 effect 函数看作是从 React 的纯函数世界通往命令式世界的应急通道。

Author: > Think of effects as an escape hatch from React's purely functional world into the imperative world
我感觉这句话的 effects 指的应该是传给 `useEffect` 的函数（文章后面的 effect 同理）。React 开了个口，让这种定义在函数组件内却包含命令式代码的函数，能通过 `useEffect` Hook 来完成不该在函数出现的副作用操作。

> 相反，应该使用 `useEffect` 来完成上述操作。传给 `useEffect` 的函数会在组件渲染完成后执行。你可以把 effect 函数看作是从 React 的纯函数世界通往命令式世界的应急通道。

Reviewer: > > Think of effects as an escape hatch from React's purely functional world into the imperative world.
> 
> 我感觉这句话的 effects 指的应该是传给 `useEffect` 的函数（文章后面的 effect 同理）。React 开了个口，让这种定义在函数组件内却包含命令式代码的函数，能通过 `useEffect` Hook 来完成不该在函数出现的副作用操作。

其实传递给 `useEffect` 的函数里做的操作就是命令式编程的内容（比如DOM操作、接口请求，它就是包含副作用的函数），也可以看做它就是副作用本身。这个副作用函数从广义上来看就属于副作用。原文里并没有特定用 effect function这个词而是只用了 effect，我觉得按照广义来理解稍微合适一点点。

> 相反，应该使用 `useEffect` 来完成上述操作。传给 `useEffect` 的函数会在组件渲染完成后执行。你可以把 effect 函数看作是从 React 的纯函数世界通往命令式世界的应急通道。

Author: 其他 Hook 文档的翻译都保留 `effect` ，此处暂时也做保留。

默认情况下，effect 将在每轮渲染结束后执行，但你可以选择让它 [在只有某些值改变的时候](#conditionally-firing-an-effect) 才执行。

> #### 清理 effect {#cleaning-up-an-effect}

Author: 保留 effect

通常，effect 内部会产生诸如订阅或计时器 ID 等资源，必须在组件卸载前清除它们。为此，传给 `useEffect` 的函数可能需要返回一个清理函数。如下，一个创建订阅的例子：

```js
useEffect(() => {
  const subscription = props.source.subscribe();
  return () => {
    // 清除 subscription
    subscription.unsubscribe();
  };
});
```

为防止内存泄漏，清理函数将在组件卸载前执行。另外，如果组件多次渲染（通常如此），**在执行下一轮的 effect 之前，上一轮的 effect 就已经被清理了**。在我们的示例中，意味着组件每一轮更新都会创建新的订阅。若想避免每次都调用 effect，请参阅下一章节。

> #### effect 的调用时机 {#timing-of-effects}

Author: 保留 effect

> #### effect 的调用时机 {#timing-of-effects}

Reviewer 2: 参考 useEffect。

> 与 `componentDidMount`、`componentDidUpdate` 不同，在浏览器完成布局与绘制**之后**，传给 `useEffect` 的函数会在一个延迟事件中被调用。这使得 `useEffect` 能跟许多常见的副作用，如订阅设置、事件处理器等搭配使用，因为大多数副作用不应该阻塞视图更新。

Author: > 因为大多数类型的工作不应阻塞浏览器对屏幕的更新

感觉有点点拗口

然而，并非所有 effect 都可以被延迟。例如，在浏览器执行下一次绘制前，用户可见的 DOM 变更就必须同步执行，这样用户才不会感觉到视觉上的不一致。（在概念上类似于被动事件监听器和主动事件监听器的区别。）React 提供了额外的 [`useLayoutEffect`](#uselayouteffect) Hook 来处理这类 effect。它和 `useEffect` 签名相同，区别只是调用时机不同。

> 浏览器绘制完成之后，`useEffect` 会确保在新的组件渲染前调用 effect。React 将在组件更新前刷新上一轮渲染的 effect。

Author: 保留 effect

> #### effect 的条件调用 {#conditionally-firing-an-effect}

Author: 保留 effect

> 默认情况下，effect 会在每轮组件渲染完成后被调用。这样一旦 effect 的依赖发生变化了，它就会被重新创建。

Author: 保留 effect

});
```

#### 跳过 state 更新 {#bailing-out-of-a-state-update}

> 假如你在调用 State Hook 的更新函数时，给它传入当前的 state 值，React 将跳过子组件的渲染及副作用的执行。（React 使用 [`Object.is` 比较算法](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/is#Description) 来比较 state。）

Author: > Suggested change 
> 如果在调用 State Hook 的更新函数时，并传入当前的 state

这里应该是：“如果调用 State Hook 的更新函数并传入当前的 state”？

> 假如你在调用 State Hook 的更新函数时，给它传入当前的 state 值，React 将跳过子组件的渲染及副作用的执行。（React 使用 [`Object.is` 比较算法](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/is#Description) 来比较 state。）

Reviewer 2: 如果调用 State Hook 的更新函数并传入当前的 state 时

> 假如你在调用 State Hook 的更新函数时，给它传入当前的 state 值，React 将跳过子组件的渲染及副作用的执行。（React 使用 [`Object.is` 比较算法](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/is#Description) 来比较 state。）

Author: ```suggestion
调用 State Hook 的更新函数并传入当前的 state 时，React 将跳过子组件的渲染及副作用的执行。（React 使用 [`Object.is` 比较算法](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/is#Description) 来比较 state。）
```

> 

Reviewer 3: specific component 

翻译成  某特定组件  是不是会贴近一点

Author: > that specific component

其实就是指“该组件”

需要注意的是，React 可能仍需要在跳过渲染前再次渲染该组件。由于 React 不会对组件树的深层节点进行不必要的渲染，所以大可不必担心。如果你在渲染期间执行了高开销的计算，则可以使用 `useMemo` 来进行优化。

### `useEffect` {#useeffect}

```js
useEffect(didUpdate);
```

（其）接收一个包含命令式、可能有副作用代码的函数。

> 在函数组件主体内（这里指在 React 渲染阶段）改变 DOM、添加订阅、设置定时器、记录日志以及执行其他包含副作用的操作都是不被允许的，因为这可能引发令人困惑的 bug 并破坏 UI 的一致性。

Author: ```suggestion
在函数组件主体内（这里指在 React 渲染阶段）改变 DOM、添加订阅、设置定时器、记录日志以及执行其他包含副作用的操作都是不被允许的，因为这可能会产生莫名其妙的 bug 并破坏 UI 的一致性。
```

> 在函数组件主体内（这里指在 React 渲染阶段）改变 DOM、添加订阅、设置定时器、记录日志以及执行其他包含副作用的操作都是不被允许的，因为这可能引发令人困惑的 bug 并破坏 UI 的一致性。

Reviewer 2: 嗯~ OK

> 在函数组件主体内（这里指在 React 渲染阶段）改变 DOM、添加订阅、设置定时器、记录日志以及执行其他包含副作用的操作都是不被允许的，因为这可能引发令人困惑的 bug 并破坏 UI 的一致性。

Reviewer 2: 别忘记折叠的部分

使用 `useEffect` 来替你完成副作用操作。传递给 `useEffect` 的函数会在组件渲染到屏幕之后才执行。你可以把 effect 看作是从 React 的纯函数式世界通往命令式世界的逃生通道。

默认情况下，effect 将在每轮渲染结束后执行，但你可以选择让它 [在只有某些值改变的时候](#conditionally-firing-an-effect) 才执行。

#### 清除 effect {#cleaning-up-an-effect}

> 通常，组件从屏幕卸载前需要将 effect 创建的诸如订阅或计时器 ID 等资源清除。要实现这一点，就需要传入 `useEffect` 的函数返回一个清除函数。以下就是一个创建订阅的例子：

Reviewer 2: ```suggestion
通常，组件卸载时需要清除 effect 创建的诸如订阅或计时器 ID 等资源清除。要实现这一点，就需要为传入 `useEffect` 的函数返回一个清除函数。以下就是一个创建订阅的例子：
```

Author: > Suggested change 
> 就需要为传入 `useEffect` 的函数返回一个清除函数

我感觉用“为...返回一个清理函数”有点怪怪的，用“让”是不是好理解一点？

```suggestion
通常，组件从屏幕卸载前需要将 effect 创建的诸如订阅或计时器 ID 等资源清除。要实现这一点，需要让传入 `useEffect` 的函数返回一个清除函数。以下就是一个创建订阅的例子：
```

Reviewer 2: 我觉得从屏幕没必要，所以去掉吧。

Reviewer 2: ```suggestion
通常，组件卸载时需要清除 effect 创建的诸如订阅或计时器 ID 等资源清除。要实现这一点，`useEffect` 函数需返回一个清除函数。以下就是一个创建订阅的例子：
```

Reviewer 2: 这样吧，不纠结让还是为了，我觉得为没有问题。

```js
useEffect(() => {
  const subscription = props.source.subscribe();
  return () => {
    // 清除订阅
    subscription.unsubscribe();
  };
});
```

> 为防止内存泄漏，清除函数将在组件从 UI 上卸载前执行。另外，如果组件多次渲染（通常如此），那么，**在执行下一个 effect 之前，前一个 effect 就已经被清除了**。在上述示例中，这意味着组件的每一次更新都会创建新的订阅。若想避免每次更新都触发 effect 的执行，请参阅下一章节。

Reviewer 2: ```suggestion
为防止内存泄漏，清除函数会在组件卸载前执行。另外，如果组件多次渲染（通常如此），则**在执行下一个 effect 之前，上一个 effect 就已被清除**。在上述示例中，意味着组件的每一次更新都会创建新的订阅。若想避免每次更新都触发 effect 的执行，请参阅下一小节。
```

#### effect 的执行时机 {#timing-of-effects}

与 `componentDidMount`、`componentDidUpdate` 不同的是，在浏览器完成布局与绘制**之后**，传给 `useEffect` 的函数会在一个延迟事件中被调用。这使得它适用于许多常见的副作用场景，例如设置订阅和事件处理程序，因为大多数类型的工作不应阻塞浏览器更新屏幕。

然而，并非所有 effect 都可以被延迟执行。例如，在浏览器执行下一次绘制前，用户可见的 DOM 变更就必须同步执行，这样用户才不会感觉到视觉上的不一致。（概念上类似于被动事件监听器和主动事件监听器的区别。）React 为此提供了一个额外的 [`useLayoutEffect`](#uselayouteffect) Hook 来处理这类 effect。它和 `useEffect` 的签名相同，区别只是调用时机不同。

> 虽然 `useEffect` 会在浏览器绘制后延迟执行，但其保证会在任何新的渲染前执行。React 将在组件更新前刷新上一轮渲染的 effect。

Reviewer 2: ```suggestion
虽然 `useEffect` 会在浏览器绘制后延迟执行，但会保证在任何新的渲染前执行。React 将在组件更新前刷新上一轮渲染的 effect。
```

);
```

现在，只有当 `props.source` 改变了才会重新创建订阅。

> 注意
>

> > 如果使用此优化，请确保依赖项数组包含**组件作用域（如 props 和 state）内被 effect 使用的所有随时间变化的值**。否则，你的代码将会引用上一轮渲染的旧数据。了解更多有关 [如何处理函数](/docs/hooks-faq.html#is-it-safe-to-omit-functions-from-the-list-of-dependencies) 和如何处理 [频繁变化的数组值](/docs/hooks-faq.html#what-can-i-do-if-my-effect-dependencies-change-too-often) 的信息。

Author: 这段跟 `使用 Effect Hook` 同步了

);
```

> 现在，只有当 `props.source` 改变后才会重新创建订阅。

Reviewer 2: ```suggestion
此时，只有当 `props.source` 改变后才会重新创建订阅。
```

>注意
>

> >如果你要使用此优化方式，请确保数组中包含了**所有外部作用域中会发生变化且在 effect 中使用的变量**，否则你的代码会引用到先前渲染中的旧变量。请参阅文档，了解更多关于 [如何处理函数](/docs/hooks-faq.html#is-it-safe-to-omit-functions-from-the-list-of-dependencies) 以及 [数组频繁变化时的措施](/docs/hooks-faq.html#what-can-i-do-if-my-effect-dependencies-change-too-often) 的内容。

Reviewer 4: ```suggestion
>如果你要使用此优化方式，请确保数组中包含了**所有外部作用域中会发生变化且在 effect 中使用的变量**，否则你的代码会引用到先前渲染中的旧变量。请参阅文档，了解更多关于[如何处理函数](/docs/hooks-faq.html#is-it-safe-to-omit-functions-from-the-list-of-dependencies) 以及[数组频繁变化时的措施](/docs/hooks-faq.html#what-can-i-do-if-my-effect-dependencies-change-too-often) 的内容。
```

> >如果你要使用此优化方式，请请确保数组中包含了**所有外部作用域中会随时间变化并且在 effect 中使用的变量**，否则你的代码会引用到先前渲染中的旧变量。参阅文档，了解更多关于 [如何处理函数](/docs/hooks-faq.html#is-it-safe-to-omit-functions-from-the-list-of-dependencies) 以及 [数组频繁变化时的措施](/docs/hooks-faq.html#what-can-i-do-if-my-effect-dependencies-change-too-often) 的内容。

Reviewer 2: ```suggestion
>如果你要使用此优化方式，请确保数组中包含了**所有外部作用域中会发生变化且在 effect 中使用的变量**，否则你的代码会引用到先前渲染中的旧变量。请参阅文档，了解更多关于 [如何处理函数](/docs/hooks-faq.html#is-it-safe-to-omit-functions-from-the-list-of-dependencies) 以及 [数组频繁变化时的措施](/docs/hooks-faq.html#what-can-i-do-if-my-effect-dependencies-change-too-often) 的内容。
```

>

> >如果你想要执行只运行一次的 effect（仅在组件挂载和卸载时执行），可以传递一个空数组（`[]`）作为第二个参数。这就告诉 React 你的 effect 不依赖于 props 或 state 中的任何值，所以它永远都不需要重复执行。这并不算是一种特殊情况 —— 依然遵循输入数组的工作方式。

Author: 这里的译文需要和[使用 Effect Hook](/reactjs/zh-hans.reactjs.org/blob/master/content/docs/hooks-effect.md#提示-通过跳过-effect-进行性能优化-tip-optimizing-performance-by-skipping-effects)中“提示”的译文统一吧？

> >如果你想要执行只运行一次的 effect（仅在组件挂载和卸载时执行），可以传递一个空数组（`[]`）作为第二个参数。这就告诉 React 你的 effect 不依赖于 props 或 state 中的任何值，所以它永远都不需要重复执行。这并不算是一种特殊情况 —— 依然遵循输入数组的工作方式。

Reviewer 2: @Yuiham 那边如果有的话，就统一下

> >如果你想要执行只运行一次的 effect（仅在组件挂载和卸载时执行），可以传递一个空数组（`[]`）作为第二个参数。这就告诉 React 你的 effect 不依赖于 props 或 state 中的任何值，所以它永远都不需要重复执行。这并不算是一种特殊情况 —— 依然遵循输入数组的工作方式。

Author: @QC-L 统一成哪个版本？

> >如果你想要执行只运行一次的 effect（仅在组件挂载和卸载时执行），可以传递一个空数组（`[]`）作为第二个参数。这就告诉 React 你的 effect 不依赖于 props 或 state 中的任何值，所以它永远都不需要重复执行。这并不算是一种特殊情况 —— 依然遵循输入数组的工作方式。

Reviewer 2: 我看看那边怎么翻译的

> >如果你想要执行只运行一次的 effect（仅在组件挂载和卸载时执行），可以传递一个空数组（`[]`）作为第二个参数。这就告诉 React 你的 effect 不依赖于 props 或 state 中的任何值，所以它永远都不需要重复执行。这并不算是一种特殊情况 —— 依然遵循输入数组的工作方式。

Reviewer 2: ```suggestion
>如果你想要执行只运行一次的 effect（仅在组件挂载和卸载时执行），你可以传递一个空数组(`[]`)作为第二个参数。这就告诉 React 你的 effect 不依赖于 props 或 state 中的任何值，所以它永远都不需要重复执行。这并不算是一种特殊情况 —— 依然遵循输入数组的工作方式。
```

> >如果你想要执行只运行一次的 effect（仅在组件挂载和卸载时执行），可以传递一个空数组（`[]`）作为第二个参数。这就告诉 React 你的 effect 不依赖于 props 或 state 中的任何值，所以它永远都不需要重复执行。这并不算是一种特殊情况 —— 依然遵循输入数组的工作方式。

Reviewer 2: 按我的版本来，那边我修改下~

>
> 如果你传了一个空数组（`[]`），effect 内部引用的 props 和 state 将始终维持它们的初始值。将 `[]` 作为第二个参数有点类似于 `componentDidMount` 和 `componentWillUnmount` 的手动模式，但通常来说有 [更好的](/docs/hooks-faq.html#is-it-safe-to-omit-functions-from-the-list-of-dependencies) [解决方案](/docs/hooks-faq.html#what-can-i-do-if-my-effect-dependencies-change-too-often) 来避免 effect 的频繁调用。另外，不要忘了 React 会把 `useEffect` 延迟到浏览器完成绘制才执行，所以多一些额外的调用也没什么问题。
>
>
> 我们推荐你把 [`exhaustive-deps`](https://github.com/facebook/react/issues/14920) 规则作为 [`eslint-plugin-react-hooks`](https://www.npmjs.com/package/eslint-plugin-react-hooks#installation) 的一部分来使用。当依赖项指定有误时，它会生成警告并提供修复建议。

依赖项数组不会作为参数传给 effect 函数。然而从概念上来说，这表示 effect 函数引用的每个值也应该出现在依赖项数组中。将来，一个足够先进的编译器可以自动创建这个数组。

### `useContext` {#usecontext}

```js
const value = useContext(MyContext);
```

> `useContext` 接收一个 context 对象（`React.createContext` 的返回值）并返回该 context 的当前值。该值由距离当前组件最近的 `<MyContext.Provider>` 的 `value` prop 确定。

Author: ```suggestion
接收一个 context 对象（`React.createContext` 的返回值）并返回该 context 的当前值。当前的 context 值由上层组件中距离当前组件最近的 `<MyContext.Provider>` 的 `value` prop 决定。
```

稍微调整了一下

> 当最近的上级 `<MyContext.Provider>` 组件更新时，该 Hook 将使用 `MyContext` provider 组件的最新 context `value` 触发组件重渲染。

Reviewer 2: 感觉这样不好

> 当最近的上级 `<MyContext.Provider>` 组件更新时，该 Hook 将使用 `MyContext` provider 组件的最新 context `value` 触发组件重渲染。

Author: 读起来有点拗口，这个长句感觉能拆短比较好理解

> 当最近的上级 `<MyContext.Provider>` 组件更新时，该 Hook 将使用 `MyContext` provider 组件的最新 context `value` 触发组件重渲染。

Reviewer 2: @Yuiham 我当时也在想拆分的问题，可以拆了

> 当最近的上级 `<MyContext.Provider>` 组件更新时，该 Hook 将使用 `MyContext` provider 组件的最新 context `value` 触发组件重渲染。

Reviewer 2: ```suggestion
当组件上层最近的 `<MyContext.Provider>` 更新时，该 Hook 会触发重渲染，并将最新的 context 的 `value` 值传递给 `MyContext` 的 provider。
```

> 当最近的上级 `<MyContext.Provider>` 组件更新时，该 Hook 将使用 `MyContext` provider 组件的最新 context `value` 触发组件重渲染。

Reviewer 2: 这样应该可以了吧？

别忘了 `useContext` 的参数必须是 *context 对象本身*：

 * **正确：** `useContext(MyContext)`
 * **错误：** `useContext(MyContext.Consumer)`
 * **错误：** `useContext(MyContext.Provider)`

调用了 `useContext` 的组件总会在 context 值变化时重新渲染。如果重渲染组件的开销较大，你可以 [通过使用 memoization 来优化](https://github.com/facebook/react/issues/15156#issuecomment-474590693)。

>提示
>

> >如果你在接触 Hook 前已经对 context API 比较熟悉，则应当明白 `useContext(MyContext)` 相当于 class 组件中的 `static contextType = MyContext` 或者 `<MyContext.Consumer>`。

Reviewer 4: 感觉差不多～～都可以吧

> >如果你在接触 Hook 前已经对 context API 比较熟悉，则应当明白 `useContext(MyContext)` 相当于 class 组件中的 `static contextType = MyContext` 或者 `<MyContext.Consumer>`。

Reviewer 2: 感觉句子太长了，断一下

> >如果你在接触 Hook 前已经对 context API 比较熟悉，则应当明白 `useContext(MyContext)` 相当于 class 组件中的 `static contextType = MyContext` 或者 `<MyContext.Consumer>`。

Author: ```suggestion
>如果你在接触 Hook 前已经对 context API 比较熟悉，那应该可以理解，`useContext(MyContext)` 相当于 class 组件中的 `static contextType = MyContext` 或者 `<MyContext.Consumer>`。
```

这么断句好一点？

> >如果你在接触 Hook 前已经对 context API 比较熟悉，则应当明白 `useContext(MyContext)` 相当于 class 组件中的 `static contextType = MyContext` 或者 `<MyContext.Consumer>`。

Reviewer 2: @Yuiham 可以

>
>`useContext(MyContext)` 只是让你能够*读取* context 的值和订阅 context 的变化。你仍然需要在上层组件树中使用 `<MyContext.Provider>` 来为下层组件*提供*该 context。

## 额外的 Hook {#additional-hooks}

以下介绍的 Hook，有些是上一章节基础 Hook 的变体，有些则仅在特定边缘情况下会用到。不用特意预先学习它们。

### `useReducer` {#usereducer}

```js
const [state, dispatch] = useReducer(reducer, initialArg, init);
```

[`useState`](#usestate) 的替代方案。它接收一个形如 `(state, action) => newState` 的 reducer，并返回当前的 state 以及与其配套的 `dispatch` 方法。（如果你熟悉 Redux 的话，你已经知道它是如何工作的了。）

> 在诸如 state 逻辑较复杂且包含多个子值，或者下一个 state 依赖于之前的 state 等场景下，通常 `useReducer` 会比 `useState` 更适用。而且，使用 `useReducer` 还能给那些会触发深更新的组件做性能优化，因为[你可以向子组件传递 `dispatch` 而不是回调函数](/docs/hooks-faq.html#how-to-avoid-passing-callbacks-down) 。

Reviewer 4: 在...场景下，中间的定于确实太长了。
这样如何？
```suggestion
在某些场景下，state 逻辑较复杂且包含多个子值，或者下一个 state 依赖于之前的 state，此时，`useReducer` 通常会比 `useState` 更适用。而且，使用 `useReducer` 还能给那些会触发深更新的组件做性能优化，因为[你可以向子组件传递 `dispatch` 而不是回调函数](/docs/hooks-faq.html#how-to-avoid-passing-callbacks-down) 。
```

或者

```suggestion
如果 state 逻辑较复杂且包含多个子值，或者下一个 state 依赖于之前的 state，通常 `useReducer` 会比 `useState` 更适用。而且，使用 `useReducer` 还能给那些会触发深更新的组件做性能优化，因为[你可以向子组件传递 `dispatch` 而不是回调函数](/docs/hooks-faq.html#how-to-avoid-passing-callbacks-down) 。
```

> 在诸如 state 逻辑较复杂且包含多个子值，或者下一个 state 依赖于之前的 state 等场景下，通常 `useReducer` 会比 `useState` 更适用。而且，使用 `useReducer` 还能给那些会触发深更新的组件做性能优化，因为[你可以向子组件传递 `dispatch` 而不是回调函数](/docs/hooks-faq.html#how-to-avoid-passing-callbacks-down) 。

Reviewer 2: 那重新组织下语言

```suggestion
在某些场景下，`useReducer` 会比 `useState` 更适用，例如 state 逻辑较复杂且包含多个子值，或者下一个 state 依赖于之前的 state 等。并且，使用 `useReducer` 还能给那些会触发深更新的组件做性能优化，因为[你可以向子组件传递 `dispatch` 而不是回调函数](/docs/hooks-faq.html#how-to-avoid-passing-callbacks-down) 。
```

用 reducer 重写 [`useState`](#usestate) 一章的计数器示例：

```js

> const initialState = {count: 0};

Author: 保留代码

> const initialState = {count: 0};

Author: 暂保留代码

  );
```

>注意
>

> >React 不使用 `state = initialState` 这一由 Redux 推广开来的参数约定。有时候初始值依赖于 props，因此需要在调用 Hook 时指定。如果你特别喜欢上述的参数约定，可以通过调用 `useReducer(reducer, undefined, reducer)` 来模拟 Redux 的行为，但我们不鼓励你这么做。

Author: React 不使用 `state = initialState` 的参数约定，跟 Redux 没有因果关系吧？并非因为这种约定由源自 Redux，所以 React 才不使用。

> >React 不使用 `state = initialState` 这一由 Redux 推广开来的参数约定。有时候初始值依赖于 props，因此需要在调用 Hook 时指定。如果你特别喜欢上述的参数约定，可以通过调用 `useReducer(reducer, undefined, reducer)` 来模拟 Redux 的行为，但我们不鼓励你这么做。

Reviewer 4: 恩，“不使用这种约定”并不是“Redux 的推广”导致的。

> >React 不使用 `state = initialState` 这一由 Redux 推广开来的参数约定。有时候初始值依赖于 props，因此需要在调用 Hook 时指定。如果你特别喜欢上述的参数约定，可以通过调用 `useReducer(reducer, undefined, reducer)` 来模拟 Redux 的行为，但我们不鼓励你这么做。

Author: 那保留原译文

#### 惰性初始化 {#lazy-initialization}

你可以选择惰性初始化 state。为此，需要把 `init` 函数作为 `useReducer` 的第三个参数。这样初始状态将被设置为 `init(initialArg)`。

使用这种方法可以把 state 的初始化计算逻辑抽离到 reducer 外面，也方便将来对重置 state 的 action 做处理：

> ```js{1-3,11-12,19,24}

Author: 暂保留代码

}
```

#### 跳过 dispatch {#bailing-out-of-a-dispatch}

如果 Reducer Hook 的返回值与当前 state 相同，React 将跳过子组件的渲染及副作用的执行。（React 使用 [`Object.is` 比较算法](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/is#Description) 来比较 state。）

> 需要注意的是，React 可能仍需要在跳过渲染前再次渲染该组件。由于 React 不会对组件树的深层节点进行不必要的渲染，所以大可不必担心。如果你在渲染期间执行了高开销的计算，则可以使用 `useMemo` 来进行优化。

Reviewer 4: 不是有个 again 嘛😂 
句尾句号～～

> 需要注意的是，React 可能仍需要在跳过渲染前再次渲染该组件。由于 React 不会对组件树的深层节点进行不必要的渲染，所以大可不必担心。如果你在渲染期间执行了高开销的计算，则可以使用 `useMemo` 来进行优化。

Reviewer 2: 为了和前文统一，就复制过来了。忽略了

);
```

返回一个 [memoized](https://en.wikipedia.org/wiki/Memoization) 回调函数。

> 把一个内联回调函数及一个依赖项数组作为参数传入 `useCallback`，它将返回这个回调函数的 memoized 版本，这个回调函数仅在其某个依赖项改变时才更新。当你把回调函数传递给经过优化的且利用引用相等性去避免非必要渲染（例如 `shouldComponentUpdate`）的子组件时，它将非常有用。

Reviewer 2: ```suggestion
把内联回调函数及依赖项数组作为参数传入 `useCallback`，它将返回该回调函数的 memoized 版本，该回调函数仅在某个依赖项改变时才会更新。当你把回调函数传递给经过优化的并使用引用相等性去避免非必要渲染（例如 `shouldComponentUpdate`）的子组件时，它将非常有用。
```

`useCallback(fn, deps)` 相当于 `useMemo(() => fn, deps)`。

> 注意
>

> > 依赖项数组不会作为参数传给回调函数。虽然从概念上来说它表现为：所有回调函数中引用的值都应该出现在依赖项数组中。将来，一个足够先进的编译器能够自动创建这个数组。

Reviewer 2: ```suggestion
> 依赖项数组不会作为参数传给回调函数。虽然从概念上来说它表现为：所有回调函数中引用的值都应该出现在依赖项数组中。未来编译器会更加智能，届时自动创建数组将成为可能。
```

>

> > 我们推荐你把 [`exhaustive-deps`](https://github.com/facebook/react/issues/14920) 规则作为 [`eslint-plugin-react-hooks`](https://www.npmjs.com/package/eslint-plugin-react-hooks#installation) 的一部分来使用。当依赖项指定有误时，它会生成警告并提供修复建议。

Author: 跟 useEffect 文档同步

### `useMemo` {#usememo}

```js
const memoizedValue = useMemo(() => computeExpensiveValue(a, b), [a, b]);
```

返回一个 [memoized](https://en.wikipedia.org/wiki/Memoization) 值。

> 把一个“创建”函数和一个依赖项数组传入 `useMemo`，它仅会在某个依赖项改变时才重新计算 memoized 值。这种优化有助于避免在每次渲染时都进行高开销的计算。

Reviewer 2: ```suggestion
把“创建”函数和依赖项数组作为参数传入 `useMemo`，它仅会在某个依赖项改变时才重新计算 memoized 值。这种优化有助于避免在每次渲染时都进行高开销的计算。
```

记住，传入 `useMemo` 的函数会在渲染期间执行。请不要在这个函数内部执行与渲染无关的操作，诸如副作用这类的操作属于 `useEffect` 的适用范畴，而不是 `useMemo`。

如果没有提供依赖项数组，`useMemo` 在每次渲染时都会计算新的值。

>  **你可以把 `useMemo` 作为性能优化的手段，但不要把它当成语义上的保证。**将来，React 可能会选择“遗忘”以前的一些 memoized 值，并在下次渲染时重新计算它们，比如为离屏组件释放内存。先编写即使在没有 `useMemo` 的情况下也可以执行的代码——之后再在你的代码中添加 `useMemo` 来优化性能。

Reviewer 2: ```suggestion
 **你可以把 `useMemo` 作为性能优化的手段，但不要把它当成语义上的保证。**将来，React 可能会选择“遗忘”以前的一些 memoized 值，并在下次渲染时重新计算它们，比如为离屏组件释放内存。先编写在没有 `useMemo` 的情况下也可以执行的代码 —— 之后再在你的代码中添加 `useMemo`，以达到优化性能的目的。
```

> 注意
>

> > 依赖项数组不会作为参数传给“创建”函数。然虽然从概念上来说它表现为：所有“创建”函数中引用的值都应该出现在依赖项数组中。将来，一个足够先进的编译器能够自动创建这个数组。

Reviewer 2: ```suggestion
> 依赖项数组不会作为参数传给“创建”函数。然虽然从概念上来说它表现为：所有“创建”函数中引用的值都应该出现在依赖项数组中。未来编译器会更加智能，届时自动创建数组将成为可能。
```

>
> 我们推荐你把 [`exhaustive-deps`](https://github.com/facebook/react/issues/14920) 规则作为 [`eslint-plugin-react-hooks`](https://www.npmjs.com/package/eslint-plugin-react-hooks#installation) 的一部分来使用。当依赖项指定有误时，它会生成警告并提供修复建议。

### `useRef` {#useref}

```js
const refContainer = useRef(initialValue);
```

`useRef` 返回一个可变的 ref 对象，其 `.current` 属性被初始化为 `useRef` 的参数（`initialValue`）。ref 对象在组件的整个生命周期内保持不变。

一个常见的用例，命令式地访问子组件：

```js
function TextInputWithFocusButton() {
  const inputEl = useRef(null);
  const onButtonClick = () => {
    // `current` 指向挂载到 DOM 上的文本输入元素
    inputEl.current.focus();
  };

>   return (

Author: 保留代码

}
```

本质上，`useRef` 就像是可以在其 `.current` 属性中保存一个可变值的“盒子”。

> 你应该熟悉 ref 这一种主要的[访问 DOM](/docs/refs-and-the-dom.html) 的方式。如果你将 ref 对象以 `<div ref={myRef} />` 形式传入组件，则无论该节点如何改变，React 都会将 ref 对象的 `.current` 属性设置为相应的 DOM 节点。

Reviewer 4: ```suggestion
你应该熟悉 ref 这一种[访问 DOM](/docs/refs-and-the-dom.html) 的主要方式。如果你将 ref 对象以 `<div ref={myRef} />` 形式传入组件，则无论该节点如何改变，React 都会将 ref 对象的 `.current` 属性设置为相应的 DOM 节点。
```

> 然而，`useRef()` 比 `ref` 属性更有用。它可以[很方便地保持任何可变值](/docs/hooks-faq.html#is-there-something-like-instance-variables)，其类似于在 class 中使用实例字段的方式。

Reviewer 2: ```suggestion
然而，`useRef()` 比 `ref` 属性更有用。它可以[很方便地保存任何可变值](/docs/hooks-faq.html#is-there-something-like-instance-variables)，其类似于在 class 中使用实例字段的方式。
```

FancyInput = forwardRef(FancyInput);
```

在本例中，渲染 `<FancyInput ref={fancyInputRef} />` 的父组件可以调用 `fancyInputRef.current.focus()`。

### `useLayoutEffect` {#uselayouteffect}

> 其函数签名与 `useEffect` 相同，但它会在所有的 DOM 变更之后同步调用 effect。可以使用它来读取 DOM 布局并同步触发重渲染。在浏览器执行绘制之前，`useLayoutEffect` 内部的更新计划将被同步刷新。

Reviewer 2: ```suggestion
其函数结构与 `useEffect` 相同，但它会在所有的 DOM 变更之后同步调用 effect。可以使用它来读取 DOM 布局并同步触发重渲染。在浏览器执行绘制之前，`useLayoutEffect` 内部的更新计划将被同步刷新。
```

Author: signature 翻译成“函数结构”？

Reviewer 4: signature 就是函数签名的意思。这是一个通用的术语。

Reviewer 2: 我当时是出于让读者好理解的角度，毕竟有人不知道函数签名。

Reviewer 2: 这里遵循你的吧，我把函数签名加到术语表。

尽可能使用标准的 `useEffect` 以避免阻塞视觉更新。

> 提示
>
> 如果你正在将代码从 class 组件迁移到使用 Hook 的函数组件，则需要注意 `useLayoutEffect` 与 `componentDidMount`、`componentDidUpdate` 的调用阶段是一样的。但是，我们推荐你**一开始先用 `useEffect`**，只有当它出问题的时再尝试使用 `useLayoutEffect`。
>

> >如果你使用服务端渲染，请记住，*无论* `useLayoutEffect` *还是* `useEffect` 都无法在 Javascript 代码加载完成之前执行。这就是为什么在服务端渲染组件中包含 `useLayoutEffect` 代码时会触发 React 告警。解决这个问题，需要将其中的代码逻辑移至 `useEffect` 中（如果首次渲染不需要这段逻辑），或是将该组件延迟到客户端渲染完成后再显示（如果直到 `useLayoutEffect` 执行之前 HTML 都显示错乱的话）。

Reviewer 2: ```suggestion
>如果你使用服务端渲染，请记住，*无论* `useLayoutEffect` *还是* `useEffect` 都无法在 Javascript 代码加载完成之前执行。这就是为什么在服务端渲染组件中引入 `useLayoutEffect` 代码时会触发 React 告警。解决这个问题，需要将代码逻辑移至 `useEffect` 中（如果首次渲染不需要这段逻辑的情况下），或是将该组件延迟到客户端渲染完成后再显示（如果直到 `useLayoutEffect` 执行之前 HTML 都显示错乱的情况下）。
```

>

> >可以通过使用 `showChild && <Child />` 进行条件渲染并结合使用 `useEffect(() => { setShowChild(true); }, [])` 延迟显示组件的方式，将需要依赖布局 effect 的组件从服务端渲染的 HTML 中移除。这样，在客户端渲染完成之前，UI 就不会像之前那样显示错乱了。

Reviewer 2: ```suggestion
>若要从服务端渲染的 HTML 中排除依赖布局 effect 的组件，可以通过使用 `showChild && <Child />` 进行条件渲染，并使用 `useEffect(() => { setShowChild(true); }, [])` 延迟展示组件。这样，在客户端渲染完成之前，UI 就不会像之前那样显示错乱了。
```

> > 为了从服务端渲染的 HTML 中移除用到 layout effect 的组件，可以通过 `showChild && <Child />` 进行条件渲染，并使用 `useEffect(() => { setShowChild(true); }, [])` 延迟显示组件。这样，在客户端渲染完成之前，UI 就不会出现坏掉的样子。

Author: > a component that needs layout effects

翻译调整为：需要依赖布局 effect 的组件。会不会好理解一些？

> > 为了从服务端渲染的 HTML 中移除用到 layout effect 的组件，可以通过 `showChild && <Child />` 进行条件渲染，并使用 `useEffect(() => { setShowChild(true); }, [])` 延迟显示组件。这样，在客户端渲染完成之前，UI 就不会出现坏掉的样子。

Reviewer 2: 可以的~

### `useDebugValue` {#usedebugvalue}

```js
useDebugValue(value)
```

> `useDebugValue` 可用于在 React 开发者工具中显示自定义 hook 标签。

Reviewer 4: 避免产生`自定义“hook”标签`这种歧义
uggestion
`useDebugValue` 可用于在 React 开发者工具中显示自定义 hook 的标签。
```

> 例如，[“编写自己的 Hook”](/docs/hooks-custom.html) 这章中描述的名为 `useFriendStatus` 的自定义 Hook：

Reviewer 2: ```suggestion
例如，[“自定义 Hook”](/docs/hooks-custom.html) 章节中描述的名为 `useFriendStatus` 的自定义 Hook：
```

```js{6-8}
function useFriendStatus(friendID) {
  const [isOnline, setIsOnline] = useState(null);

  // ...

  // 在开发者工具中的这个 Hook 旁边显示标签

>   // e.g. "FriendStatus: Online"

Author: 保留代码

>   useDebugValue(isOnline ? 'Online' : 'Offline');

Author: 保留代码
